### PR TITLE
Support option to skip the optimizer for training step

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

This PR introduces a mechanism to skip training steps during severe loss or gradient anomalies (b/489540436). Reference implementation at [OLMo-core](https://github.com/allenai/OLMo-core/blob/c757b7c3c15197154c753d883330afbfa4869dcc/src/olmo_core/optim/skip_step_optimizer.py#L12).
* Add configs in `base.yml` & `types.py`
* Implemented `skip_step_on_spikes` as an `optax.GradientTransformationExtraArgs` wrapper
* Training loop integration for optimizer updates

# Tests

* Add a unit test in `tests/unit/optimizers_test.py`
* End-to-end training functional comparison: [link](https://screenshot.googleplex.com/6Ap53eoNChKHo66)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/optimizers/optimizers.py [hunk 1]
```diff
@@ -48,6 +48,115 @@ def get_adamw_mask(config):
...
+        "count": jnp.zeros((), dtype=jnp.int32),
+    }
+
+  def update_fn(updates, state, params=None, **extra_args):
+    loss = extra_args.get("loss", None)
+    grad_norm = extra_args.get("grad_norm", None)
+
+    # Fallback to standard update if loss is not provided
+    if loss is None:
+      inner_updates, new_inner_state = inner_opt.update(updates, state["inner_state"], params, **extra_args)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/maxtext/optimizers/optimizers.py:108]: Make sense, and will update.

File: src/maxtext/optimizers/optimizers.py [hunk 2]
```diff
@@ -48,6 +48,115 @@ def get_adamw_mask(config):
...
+    if grad_norm is not None:
+      is_grad_norm_ok = (grad_norm - grad_norm_mean) <= scaling_factor * grad_norm_std
+      is_ok = jnp.logical_and(is_loss_ok, is_grad_norm_ok)
+    else:
+      is_ok = is_loss_ok
+
+    # Only enforce skip if we have at least half the interval filled (or 2 elements minimum)
+    min_history = max(2, interval // 2)
+    is_warmup = (count + 1) < min_history
+    is_ok = jnp.logical_or(is_warmup, is_ok)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/maxtext/optimizers/optimizers.py:136]: I'd like to stick to the reference implementation for now. It looks likes PyTorch skips the NaN step but stays "stuck" in a skip-loop for the next ~128 steps.

File: src/maxtext/optimizers/optimizers.py [hunk 3]
```diff
@@ -48,6 +48,115 @@ def get_adamw_mask(config):
...
+      return inner_updates, state["inner_state"]
+
+    inner_updates, new_inner_state = jax.lax.cond(is_ok, do_update, skip_update)
+
+    # Update rolling buffers (we append even if skipped so spikes can become the new baseline)
+    idx = count % interval
+    new_losses = losses.at[idx].set(loss)
+    new_grad_norms = grad_norms
+    if grad_norm is not None:
+      new_grad_norms = grad_norms.at[idx].set(grad_norm)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/maxtext/optimizers/optimizers.py]: Similar above.

File: src/maxtext/optimizers/optimizers.py [hunk 4]
```diff
@@ -48,6 +48,119 @@ def get_adamw_mask(config):
...
+    grad_norm = extra_args.pop("grad_norm", None)
+
+    # Fallback to standard update if loss is not provided
+    if loss is None:
+      inner_updates, new_inner_state = inner_opt.update(updates, state["inner_state"], params, **extra_args)
+      return inner_updates, {
+          "inner_state": new_inner_state,
+          "losses": state["losses"],
+          "grad_norms": state["grad_norms"],
+          "count": state["count"],
```
Comment (gagika) [inline src/maxtext/optimizers/optimizers.py:113]: could you also add `is_skipped` in the metrics?

new_state = {
    "inner_state": new_inner_state,
    "losses": new_losses,
    "grad_norms": new_grad_norms,
    "count": count + 1,
    "is_skipped": jnp.logical_not(is_ok) 
}

**Review by gagika — APPROVED** [review]
thanks, one minor comment

File: src/maxtext/optimizers/optimizers.py (hunk 4 above)
Comment (RissyRan) [reply src/maxtext/optimizers/optimizers.py:113]: Added, and also updated the unit tests accordingly.

File: src/maxtext/optimizers/optimizers.py [hunk 5]
```diff
@@ -48,6 +48,122 @@ def get_adamw_mask(config):
...
+          "is_skipped": jnp.array(False, dtype=jnp.bool_),
```
Comment (aireenmei) [inline src/maxtext/optimizers/optimizers.py:114]: Will user see is_skipped in logging? It would be good if the logging shows "skipped" when the step is skipped, but no need to show is_skipped for every step since the majority are non-skipped. Alternatively we can show a separate warning message when a step is skipped.

Comment (RissyRan) [reply src/maxtext/optimizers/optimizers.py:114]: Good call! Let me add a log when skip = True.
